### PR TITLE
Fix invalid state

### PR DESCRIPTION
**Pull request by willmcgugan**
Fixes https://github.com/Textualize/textual/issues/4884

### CODE REVIEW

File: tests/test_screens.py [hunk 1]
```diff
@@ -537,3 +537,52 @@ def get_default_screen(self) -> Screen:
         assert len(app.screen_stack) == 1
         assert isinstance(app.screen_stack[0], CustomScreen)
         assert app.screen is app.screen_stack[0]
+
+
+async def test_worker_cancellation():
```
Comment (darrenburns) [inline tests/test_screens.py:542]: I'm not really following this test or how it's related to this issue to be honest 😅 It doesn't really describe the issue and how it's testing it/ensuring it's fixed. Could you add comments to it to show where the issue occurred?

**Review by darrenburns — CHANGES_REQUESTED** [review]
I'm unsure how this relates to the linked issue and have requested some extra info in the test to describe what's broken.

File: src/textual/app.py [hunk 2]
```diff
@@ -2181,7 +2181,7 @@ async def push_screen_wait(
             The screen's result.
         """
         await self._flush_next_callbacks()
-        return await self.push_screen(screen, wait_for_dismiss=True)
+        return await asyncio.shield(self.push_screen(screen, wait_for_dismiss=True))
```
Comment (darrenburns) [inline src/textual/app.py:2184]: Maybe a comment on _why_ we're shielding here/the circumstances that the shield protects against?

**Comment by willmcgugan** [comment]
@darrenburns Added some comments.

It's a little tricky to describe. Basically without the fix, cancelling a worker waiting on a screen produced some undefined behaviour.

**Review by darrenburns — APPROVED** [review]

**Pull request merged**